### PR TITLE
feat: build docker image for e2e testing

### LINKED ISSUE #21280: Push new docker image on every commit to the master branch
In order to seamlessly integrate the UI assets from the `ui` repository with OSS, we need a way to run the end-to-end tests between the two. When changes are made in `ui`, the tests need to be run against the latest OSS backend, and when changes are made to OSS, the tests need to be run against the latest `ui` frontend. The OSS changes triggering tests flow is described in [ui#1007](https://github.com/influxdata/ui/issues/1007). This issue is related to optimizing what happens when changes to the UI trigger OSS backend tests.

The `ui` repository accomplishes this using a docker image of the backend and running the end-to-end tests against that. Ideally, this docker image would represent the *latest* version of the `master` branch in `influxdb`.

Currently, a new docker image is pushed from the master branch every night. This provides a reasonably up-to-date view of the OSS backend, but it would be better if a new image was pushed on every change to the `master` branch, that way the `ui` repository could be confident it was always testing against the latest OSS backend. It may be useful to push this image to a separate repo on the image host dedicated to these testing images.

### PR DESCRIPTION
**Pull request by wbaker85**
Closes #21280

The main change here is updating the CircleCI config to build and push an image that can be used for UI e2e testing. With this change, the CI will build an image using the binary generated in the build step and the existing `Dockerfile`. It will then use the image with our *existing* e2e tests. The image will be pushed to quay as `quay.io/influxdb/oss-acceptance:latest` (which will override whatever the previous `latest` was, keeping this up-to-date) and also `quay.io/influxdb/oss-acceptance:COMMIT_SHA` for reference.

This `latest` image will allow the e2e tests initiated from changes in the `ui` repository to always use the latest OSS backend in the `monitor-ci` tests. Eventually, we will use the `quay.io/influxdb/oss-acceptance:COMMIT_SHA` image to replace the existing e2e step with the `monitor-ci` pipeline as well. In that case, the `COMMIT_SHA` tagged image will be pushed to quay prior to the e2e step so that it is available to the external pipeline.

Other than that, there is some cleanup and consolidation of the `jsdeps` job and things related to `yarn`. We only need anything to do with `yarn` in the e2e job now, which installs cypress to run our legacy tests. Once we move over to the `monitor-ci` pipeline we will be able to remove this final trace of javascript from the CI. For now, I've isolated it all in one place: The e2e job.

Running the e2e tests with a docker image had some interesting side effects. I changed the executor to a VM, since running it as docker-in-docker was problematic. This combination of OSS-in-docker and VM makes the tests take a little longer to run. Also, it causes video generation of the tests to actually work! - But this makes the tests take even longer. Because of this, I turned off video generation. When we move over to running the `monitor-ci` pipeline for e2e testing we will get videos as artifacts again, since that pipeline has more optimization around running the tests in parallel to make the process take less time. Finally, running the tests like this caused one of them to fail quite consistently. I've skipped this test for now, which is consistent with the testing from the `ui` repository, [which also skips the equivalent test code](https://github.com/influxdata/ui/blob/a131cdbe9cd541634b0e97ad54ce35ff098ee62c/cypress/e2e/shared/explorer.test.ts#L440).

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -47,6 +44,13 @@ workflows:
           filters:
             branches:
               only: master
+      - share-testing-image:
+          filters:
+            branches:
+              only:
+                - master
+          requires:
+            - e2e
```
Comment (wbaker85) [inline .circleci/config.yml:53]: By the time this is merging to master, I'm assuming all the tests are good, and just doing one final e2e test to account for an unlikely race condition between merges on master of `ui` vs. `influxdb`.

File: .circleci/config.yml [hunk 2]
```diff
@@ -531,6 +501,15 @@ jobs:
             - /tmp/go-cache
       - store_artifacts:
           path: bin/linux
+      - setup_remote_docker:
+          version: 19.03.13
+          docker_layer_caching: true
```
Comment (danxmoran) [inline .circleci/config.yml:506]: AFAIK this only affects the `machine` executor.

File: .circleci/config.yml [hunk 3]
```diff
@@ -531,6 +501,15 @@ jobs:
...
+      - run:
+          name: Build the candidate docker image
+          command: |
+            cp ./bin/linux/{influx,influxd} .
+            docker build -f docker/influxd/Dockerfile -t quay.io/influxdb/oss-acceptance:${CIRCLE_SHA1} .
+            docker save -o docker-image.tar quay.io/influxdb/oss-acceptance:${CIRCLE_SHA1}
```
Comment (danxmoran) [inline .circleci/config.yml:512]: TIL!

File: .circleci/config.yml [hunk 4]
```diff
@@ -531,6 +501,15 @@ jobs:
             - /tmp/go-cache
       - store_artifacts:
           path: bin/linux
+      - setup_remote_docker:
+          version: 19.03.13
+          docker_layer_caching: true
+      - run:
+          name: Build the candidate docker image
```
Comment (danxmoran) [inline .circleci/config.yml:508]: Do we want to skip this step when building from a fork? Does it add any significant runtime to the build stage?

Comment (danxmoran) [reply .circleci/config.yml:508]: Alt: since we're already publishing the `influx` and `influxd` binaries as artifacts from this stage, could we put the Docker-build logic in the job that runs the monitor-ci tests?

File: .circleci/config.yml [hunk 5]
```diff
@@ -736,3 +727,21 @@ jobs:
...
+    steps:
+      - attach_workspace:
+          at: /home/circleci/go/src/github.com/influxdata/influxdb
+      - run:
+          name: Log into Quay
+          command: docker login -u "$QUAY_USER" -p $QUAY_PASS quay.io
+      - run:
+          name: Push the image to Quay
+          command: |
+            docker load < docker-image.tar
```
Comment (danxmoran) [inline .circleci/config.yml:744]: Following up from comments on `build`: The `docker build` steps could run here. IMO the `e2e` job could continue to test the un-dockerized binary for as long as we keep it around.

File: ui/cypress.json [hunk 6]
```diff
@@ -1,4 +1,5 @@
 {
   "integrationFolder": "cypress/e2e",
-  "defaultCommandTimeout": 30000
+  "defaultCommandTimeout": 30000,
+  "video": false
```
Comment (danxmoran) [inline ui/cypress.json:4]: Why this change?

File: ui/cypress/e2e/explorer.test.ts [hunk 7]
```diff
@@ -438,7 +438,7 @@ describe('DataExplorer', () => {
       })
     })
 
-    it('shows flux signatures', () => {
+    it.skip('shows flux signatures', () => {
```
Comment (danxmoran) [inline ui/cypress/e2e/explorer.test.ts:441]: Flaky test?

File: .circleci/config.yml (hunk 1 above)
Comment (danxmoran) [reply .circleci/config.yml:53]: At what point does `e2e` go away? Once we cut over to triggering the workflow in `monitor-ci`?

File: ui/cypress.json (hunk 6 above)
Comment (wbaker85) [reply ui/cypress.json:4]: The last paragraph in my (very long) PR description goes into it in more detail, but basically running the e2e in a VM made the tests take longer and also made the video encoding actually work...which made them take even longer. At this point we've not had videos working on the e2e tests for so long that I didn't think it was worth turning them on and extending the duration of the tests. When we move to `monitor-ci` there's a bunch of optimization for running the tests in parallel and videos will work great.

File: ui/cypress/e2e/explorer.test.ts (hunk 7 above)
Comment (wbaker85) [reply ui/cypress/e2e/explorer.test.ts:441]: Yeah, also described in more detail in the PR description. For some reason running the tests in a VM with a docker container made this test fail a lot - [the equivalent test code](https://github.com/influxdata/ui/blob/a131cdbe9cd541634b0e97ad54ce35ff098ee62c/cypress/e2e/shared/explorer.test.ts#L440) is skipped in the `ui` repo as well.

File: .circleci/config.yml (hunk 1 above)
Comment (wbaker85) [reply .circleci/config.yml:53]: Yeah. We'll probably called it something more like `run-monitor-ci-tests` and it will just drop into where we are doing `e2e` now. Merges into `master` would still be required to complete `run-monitor-ci-tests` before they publish the `latest` image just to make sure everything is good.

File: .circleci/config.yml (hunk 4 above)
Comment (wbaker85) [reply .circleci/config.yml:508]: Since the binary is already built it goes pretty fast - takes about a minute. That said, every little bit helps, so it would make sense to skip the step for forks. That is, when we start using the `monitor-ci` pipeline, because right now even forks would need it since they'll use it during the `e2e` job.

Initially I had actually put the image building stuff in the `e2e` job, but then I got to thinking you'd end up rebuilding the docker image any time you needed to re-run the `e2e` job - which is more often than the `build` job. I think this will continue to be the case with the `monitor-ci` pipeline...e2e tests are finicky. So it would have the potential to add more time than needed in the case of re-running the e2e tests.

After that I actually tried building the docker image as its own separate job. That works fine but there's a bit of overhead from spinning up the environment to build it in. So I ended up here...the binary is right there, so just build it into the image.

File: .circleci/config.yml (hunk 5 above)
Comment (wbaker85) [reply .circleci/config.yml:744]: So running the `e2e` job with the docker image is largely ceremonial and actually caused some additional complications described elsewhere, but I was thinking it would be good to do it this way to have some confidence that the docker image actually works with the UI before unleashing it into the world and having the `ui` repo run its CI against it. Fundamentally if the binary works then the docker image should work too...so maybe I'm being a little too paranoid.

Comment (danxmoran) [reply .circleci/config.yml:744]: Makes sense to me, I guess it did flush out some problems like the video and flaky test. Good to leave dockerized now that it's working IMO 👍

File: .circleci/config.yml (hunk 2 above)
Comment (wbaker85) [reply .circleci/config.yml:506]: I read up a little bit on `docker_layer_caching` (DLC!) and it looks it works with both the `machine` executor and in a `setup_remote_docker`: https://circleci.com/docs/2.0/docker-layer-caching/

There's a point made at that link under *Limitations* which gets at the fact that DLC doesn't working for spinning up a container used as an environment.

I certainly didn't go as far as testing the build time with or without DLC though, so I'm not sure how much difference it actually makes.

**Review by danxmoran — APPROVED** [review]

File: .circleci/config.yml [hunk 8]
```diff
@@ -637,7 +609,26 @@ jobs:
             - yarn-deps-lock-{{ checksum "ui/yarn.lock" }}
       - run: sudo apt-get update && sudo apt-get install netcat-openbsd
       - run:
-          command: ./bin/linux/influxd --store=memory --e2e-testing=true
+          name: Install Dependencies
+          command: |
+            set +e
```
Comment (codyshepherd) [inline .circleci/config.yml:614]: Why is `e` being turned off here?

**Review by codyshepherd — APPROVED** [review]
This seems great to me, though I had one non-blocking question, inline.

Side note: as we get to understand Circle better, we may want to review where we are and aren't using docker layer caching, and imo we also want to think about standardizing the machine and docker images we use across our automation in order to firm up our support/debugging standardization, but these are more just ideas to noodle on for now.

File: .circleci/config.yml (hunk 8 above)
Comment (wbaker85) [reply .circleci/config.yml:614]: Beyond that I just copied it verbatim from [where it was before](https://github.com/influxdata/influxdb/blob/e28377f05cc92cadee11029e4b9046aaedfd32b6/.circleci/config.yml#L297) I couldn't say to tell you truth. Doesn't seem like it would be strictly necessary but I also didn't try without. Soon this whole `e2e` job will be revamped and replaced with the `monitor-ci` pipeline so it should be out the picture then at least!
